src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,86 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
 
     // Make sure that EAX is reported as live GC-ref so that any GC that kicks in while
     // executing GS cookie check will not collect the object pointed to by EAX.
-    if (!pushReg && (compiler->info.compRetType == TYP_REF))
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    //
+    // For Amd64 System V, a two-register-returned struct could be returned in RAX and RDX
+    // In such case make sure that the correct GC-ness of RDX is reported as well, so
+    // a GC object pointed by RDX will not be collected.
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (compiler->compMethodReturnsMultiRegRetType())
+        {
+            // Get the return tye of the struct.
+            SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR retStructDesc;
+            compiler->eeGetSystemVAmd64PassStructInRegisterDescriptor(compiler->info.compMethodInfo->args.retTypeClass, &retStructDesc);
+
+            assert(retStructDesc.passedInRegisters);
+
+            // In case the return type is a two-register-return, the native return type should be a struct
+            assert(varTypeIsStruct(compiler->info.compRetNativeType));
```

Comment:
I don't think this assert is adding any value because this is one of the condition checked by compMethodReturnsMultiRegRetType();